src/ops/cast.cc

```diff
@@ -85,6 +79,11 @@ Cast::Cast(FFModel &model,
          0 /*weights*/,
          1 /*outputs*/,
          input) {
+  
+  // assert is valid
+  assert (input->check_valid());
+  assert (input->dims[input->num_dims - 1].degree == 1);
```

Comment:
Adding some explanations as to what these checks are would be nice. 

Also, I wonder if we should be `assert`ing here or doing some actual error reporting. Personally not sure and it doesn't have to be solved in this PR, just wondering

Reply:
I think the syntax we sometimes use is `assert(... & "this is the error message") so at least the error message appears in the assertion, though of course an actual error reporting mechanism would be nice. You could also maybe use logging support? I don't know, maybe @jiazhihao has a clear idea of what to do for this

Reply:
For all the operators, I've just converted all the statements from `is_valid()` to asserts in the constructor. I think this is ok for now because it has the same functionality as before. As for the actual checks, it's the same issue as in the previous PR -- not clear why they are there. We punted on the decision to change any of those, but maybe we can look at each operator closely in this PR.